src/object.rs

```diff
@@ -386,6 +386,30 @@ glib_wrapper! {
     }
 }
 
+#[doc(hidden)]
+#[macro_export]
+macro_rules! glib_weak_impl {
+    ($name:ident) => {
+        #[doc(hidden)]
+        impl $crate::clone::Downgrade for $name {
+            type Target = $crate::object::WeakRef<Self>;
+
+            fn downgrade(&self) -> Self::Target {
+                <Self as $crate::object::ObjectExt>::downgrade(&self)
+            }
+        }
+
+        #[doc(hidden)]
+        impl $crate::clone::Upgrade for $crate::object::WeakRef<$name> {
```

Comment:
@antoyo How did this compile for you? We can't implement this outside of the glib trait because `WeakRef` and `Upgrade` are both declared in the glib trait. So compiling e.g. gtk with this change fails :)

I tried moving everything into a single trait but that causes some other problems.

Reply:
Fixed